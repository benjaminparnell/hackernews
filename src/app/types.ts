export type StoryType = 'new' | 'top';

export type Post = {
  title: string;
  by: string;
  url: string;
  score: number;
  descendants: number;
};
