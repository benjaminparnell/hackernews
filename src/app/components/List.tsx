import * as React from 'react';
import { parse } from 'url';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import Points from './Points';
import { Post } from '../types';

type Props = {
  posts: Post[];
};

export const ListCardLink = styled.a`
  color: ${props => props.theme.primary};

  &:visited {
    color: ${props => props.theme.secondary};
  }
`;

const ListCard = ({
  id,
  index,
  title,
  by,
  url,
  score,
  descendants
}: Post & { index: number }) => (
  <div>
    <h3>
      {index}.{' '}
      <ListCardLink href={url} target="_blank">
        {title}
      </ListCardLink>{' '}
      {url && `(${parse(url).host})`}
    </h3>
    <p>
      <Points score={score} /> by {by} |{' '}
      <Link to={`/item/${id}`}>
        {descendants ? `${descendants} comments` : 'discuss'}
      </Link>
    </p>
  </div>
);

const List = ({ posts }: Props) => (
  <div>
    {posts.map((post, index) => (
      <ListCard key={index} {...post} index={index + 1} />
    ))}
  </div>
);

export default List;
