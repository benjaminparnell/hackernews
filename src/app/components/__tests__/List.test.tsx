import * as React from 'react';
import { render } from 'react-testing-library';
import { MemoryRouter } from 'react-router-dom';
import List, { ListCard } from '../List';
import { Post } from '../../types';

const examplePost: Post = {
  id: 123,
  title: 'This is a title',
  text: 'This is a post',
  by: 'ben',
  url: 'https://google.com',
  score: 100,
  descendants: 0,
  kids: [],
  time: new Date().getTime()
};

const posts: Post[] = [examplePost, examplePost, examplePost];

describe('<List />', () => {
  const renderList = (posts: Post[]) => {
    return render(
      <MemoryRouter>
        <List posts={posts} />
      </MemoryRouter>
    );
  };

  it('should render a list of <ListCard />', () => {
    const { getAllByText } = renderList(posts);
    expect(getAllByText('This is a title').length).toBe(3);
  });
});

describe('<ListCard />', () => {
  const renderCard = (post: Post) => {
    return render(
      <MemoryRouter>
        <ListCard {...post} index={1} />
      </MemoryRouter>
    );
  };

  it('should match snapshot', () => {
    expect(renderCard(examplePost)).toMatchSnapshot();
  });

  it('should render "discuss" if its post has no descendants', () => {
    const { getByText } = renderCard(examplePost);
    getByText('discuss');
  });

  it('should render "{n} comments" if its post has n comments', () => {
    const { getByText } = renderCard({ ...examplePost, descendants: 10 });
    getByText('10 comments');
  });

  it('should strip the displayed external url down to its host', () => {
    const { container } = renderCard(examplePost);
    expect(container.innerHTML).toContain('(google.com)');
  });
});
