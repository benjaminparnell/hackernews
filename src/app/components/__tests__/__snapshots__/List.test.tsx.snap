// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListCard /> should match snapshot 1`] = `
<div>
  <h3>
    1
    .
     
    <a
      class="sc-bdVaJa dnpXWZ"
      href="https://google.com"
      target="_blank"
    >
      This is a title
    </a>
     
    (google.com)
  </h3>
  <p>
    $
    100
     points
     by 
    ben
     |
     
    <a
      href="/item/123"
    >
      discuss
    </a>
  </p>
</div>
`;
